diff --git a/CMakeLists.txt b/CMakeLists.txt
index a6cd67b..cdb87d0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -216,8 +216,9 @@ if(WITH_RUST)
     list(APPEND ALL_INC ${INC_DIR}/chewing_rs.h)
 endif()
 
-add_subdirectory(doc)
+if (NOT EMSCRIPTEN)
 add_subdirectory(data)
+endif()
 if(BUILD_TESTING)
     add_subdirectory(tests)
 endif()
@@ -229,13 +230,13 @@ add_library(common OBJECT
     src/porting_layer/src/asprintf.c
 )
 target_compile_definitions(common PRIVATE
-    CHEWING_DATADIR=\"${CMAKE_INSTALL_FULL_DATADIR}/libchewing\"
+    CHEWING_DATADIR=\"./libchewing\"
 )
 
 add_library(libchewing ${ALL_INC} src/chewing.c)
 set_target_properties(libchewing PROPERTIES LINKER_LANGUAGE C)
 target_compile_definitions(libchewing PRIVATE
-    CHEWING_DATADIR=\"${CMAKE_INSTALL_FULL_DATADIR}/libchewing\"
+    CHEWING_DATADIR=\"./libchewing\"
 )
 if(NOT WITH_RUST)
     target_sources(common PRIVATE
