diff --git a/CMakeLists.txt b/CMakeLists.txt
index ac03e10..8ce2852 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -78,7 +78,17 @@ corrosion_import_crate(MANIFEST_PATH Cargo.toml CRATES chewing_capi)
 if(BUILD_TESTING)
     corrosion_import_crate(MANIFEST_PATH Cargo.toml CRATES chewing_testhelper)
 endif()
+if (EMSCRIPTEN)
+    corrosion_add_target_rustflags(chewing_capi -C panic=unwind)
+    set(CARGO_FEATURES "")
+else()
 corrosion_import_crate(MANIFEST_PATH Cargo.toml CRATES chewing-cli)
+    set(CARGO_FEATURES "--features=threading")
+endif()
+set_target_properties(chewing_capi PROPERTIES
+    CORROSION_ADD_ARGS "${CARGO_FEATURES}"
+)
+corrosion_add_target_rustflags(chewing_capi "--remap-path-prefix=${PROJECT_SOURCE_DIR}=.")
 
 if(WITH_SQLITE3)
     corrosion_set_features(chewing_capi FEATURES sqlite)
@@ -243,7 +253,7 @@ endif()
 install(FILES ${ALL_INC} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/chewing)
 install(FILES ${PROJECT_BINARY_DIR}/chewing.pc
     DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
-install(TARGETS libchewing
+install(TARGETS chewing_capi
     EXPORT libchewingTargets
     ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
     INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
@@ -262,7 +272,6 @@ if(BUILD_DATA)
     )
 endif()
 
-install(IMPORTED_RUNTIME_ARTIFACTS chewing-cli DESTINATION ${CMAKE_INSTALL_BINDIR})
 
 # generate CMake Config files
 include(CMakePackageConfigHelpers)
diff --git a/Cargo.toml b/Cargo.toml
index 5197bad..622c1ff 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -29,6 +29,7 @@ tempfile = { workspace = true }
 default = []
 sqlite = ["dep:rusqlite"]
 sqlite-bundled = ["sqlite", "rusqlite/bundled"]
+threading = []
 
 [workspace]
 members = ["capi", "fuzzer", "tests/testhelper", "tools"]
@@ -44,7 +45,6 @@ lto = true
 opt-level = 3
 panic = "abort"
 codegen-units = 1
-debug = true
 
 [package.metadata.docs.rs]
 features = ["sqlite"]
diff --git a/capi/Cargo.toml b/capi/Cargo.toml
index e87d416..cfb932e 100644
--- a/capi/Cargo.toml
+++ b/capi/Cargo.toml
@@ -7,7 +7,7 @@ rust-version = "1.88"
 edition = "2024"
 
 [lib]
-crate-type = ["rlib", "staticlib"]
+crate-type = ["staticlib"]
 
 [dependencies]
 chewing = { version = "0.12.0", path = ".." }
diff --git a/src/dictionary/trie_buf.rs b/src/dictionary/trie_buf.rs
index 722df3f..b6ef2e7 100644
--- a/src/dictionary/trie_buf.rs
+++ b/src/dictionary/trie_buf.rs
@@ -231,6 +231,7 @@ impl TrieBuf {
     }
 
     pub(crate) fn wait(&mut self) {
+    #[cfg(feature = "threading")] {
         if let Some(join_handle) = self.join_handle.take() {
             match join_handle.join() {
                 Ok(Err(error)) => {
@@ -248,9 +249,11 @@ impl TrieBuf {
             }
         }
     }
+    }
 
     pub(crate) fn sync(&mut self) -> Result<(), UpdateDictionaryError> {
         info!("Synchronize dictionary from disk...");
+    #[cfg(feature = "threading")] {
         let make_error = |e| UpdateDictionaryError {
             message: "synchornize dictionary from disk failed",
             source: Some(Box::new(e)),
@@ -293,11 +296,12 @@ impl TrieBuf {
                 self.trie = Some(trie);
             }
         }
+    }
         Ok(())
     }
 
     pub(crate) fn checkpoint(&mut self) {
-        let make_error = |e| UpdateDictionaryError {
+        let make_error = |e: std::io::Error| UpdateDictionaryError {
             message: "failed to save snapshot",
             source: Some(Box::new(e)),
         };
@@ -318,6 +322,7 @@ impl TrieBuf {
             dirty: false,
             usage: self.usage,
         };
+    #[cfg(feature = "threading")] {
         self.join_handle = Some(thread::spawn(move || {
             let mut builder = TrieBuilder::new();
             info!("Saving snapshot...");
@@ -340,6 +345,30 @@ impl TrieBuf {
             info!("    Done");
             Ok(())
         }));
+    }
+        #[cfg(not(feature = "threading"))] {
+            let mut builder = TrieBuilder::new();
+            info!("Saving snapshot...");
+            let _ = builder.set_info(DictionaryInfo {
+                software: software_version(),
+                ..snapshot.about()
+            });
+            for (syllables, phrase) in snapshot.entries() {
+                builder.insert(&syllables, phrase);
+            }
+            info!("Flushing snapshot...");
+            builder.build(snapshot.path().unwrap());
+            info!("    Done");
+            info!("Reloading...");
+            let trie = Trie::open(self.path().unwrap());
+            if let Ok(value) = trie {
+                self.trie = Some(value)
+            }
+            if !self.dirty {
+                self.btree.clear();
+                self.graveyard.clear();
+            }
+        }
         self.dirty = false;
     }
 }
