diff --git a/CMakeLists.txt b/CMakeLists.txt
index a6cd67b..4c05c90 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,7 +11,6 @@ include(CTest)
 set(CTEST_PARALLEL_LEVEL 1)
 
 if(UNIX)
-    set(CMAKE_C_FLAGS "-g -O2 -Wall -fPIC ${CMAKE_C_FLAGS}")
     add_compile_definitions(UNDER_POSIX PIC)
 endif()
 
@@ -216,8 +215,9 @@ if(WITH_RUST)
     list(APPEND ALL_INC ${INC_DIR}/chewing_rs.h)
 endif()
 
-add_subdirectory(doc)
+if (NOT EMSCRIPTEN)
 add_subdirectory(data)
+endif()
 if(BUILD_TESTING)
     add_subdirectory(tests)
 endif()
@@ -229,13 +229,13 @@ add_library(common OBJECT
     src/porting_layer/src/asprintf.c
 )
 target_compile_definitions(common PRIVATE
-    CHEWING_DATADIR=\"${CMAKE_INSTALL_FULL_DATADIR}/libchewing\"
+    CHEWING_DATADIR=\"./libchewing\"
 )
 
 add_library(libchewing ${ALL_INC} src/chewing.c)
 set_target_properties(libchewing PROPERTIES LINKER_LANGUAGE C)
 target_compile_definitions(libchewing PRIVATE
-    CHEWING_DATADIR=\"${CMAKE_INSTALL_FULL_DATADIR}/libchewing\"
+    CHEWING_DATADIR=\"./libchewing\"
 )
 if(NOT WITH_RUST)
     target_sources(common PRIVATE
