diff --git a/meson.build b/meson.build
index cfe8d949..ea13c9cb 100644
--- a/meson.build
+++ b/meson.build
@@ -737,6 +737,7 @@ else
 endif
 
 
+'''
 # xkeyboard-config "verifier"
 xkct_config = configuration_data()
 xkct_config.set('MESON_BUILD_ROOT', meson.current_build_dir())
@@ -1247,6 +1248,7 @@ if get_option('enable-x11')
       env: bench_env,
   )
 endif
+'''
 
 
 # Documentation.
@@ -1374,6 +1376,6 @@ if meson.version().version_compare('>=0.62.0')
       'variant': get_option('default-variant'),
     }, section: 'Defaults')
     summary({
-      'merge-modes': has_merge_modes_tests,
+    #   'merge-modes': has_merge_modes_tests,
     }, section: 'Tests')
 endif
