diff --git a/Configurations/10-main.conf b/Configurations/10-main.conf
index cba57b4127..bab71b2d93 100644
--- a/Configurations/10-main.conf
+++ b/Configurations/10-main.conf
@@ -907,6 +907,14 @@ my %targets = (
         perlasm_scheme   => "elf32",
         multilib         => "x32",
     },
+    "ohos-aarch64" => {
+        inherit_from     => [ "linux-aarch64" ],
+        shared_extension => ".so"
+    },
+    "ohos-x86_64" => {
+        inherit_from     => [ "linux-x86_64" ],
+        shared_extension => ".so"
+    },
 
     "linux-ia64" => {
         inherit_from     => [ "linux-generic64" ],
diff --git a/crypto/build.info b/crypto/build.info
index 2642d30754..920935b0a0 100644
--- a/crypto/build.info
+++ b/crypto/build.info
@@ -115,7 +115,7 @@ DEFINE[../libcrypto]=$UPLINKDEF
 
 DEPEND[info.o]=buildinf.h
 DEPEND[cversion.o]=buildinf.h
-GENERATE[buildinf.h]=../util/mkbuildinf.pl "$(CC) $(LIB_CFLAGS) $(CPPFLAGS_Q)" "$(PLATFORM)"
+GENERATE[buildinf.h]=../util/mkbuildinf.pl "$(LIB_CFLAGS) $(CPPFLAGS_Q)" "$(PLATFORM)"
 
 GENERATE[uplink-x86.S]=../ms/uplink-x86.pl
 GENERATE[uplink-x86_64.s]=../ms/uplink-x86_64.pl
